sw/host/ot_certs/src/cwt.rs

```diff
@@ -0,0 +1,354 @@
+
+use anyhow::{anyhow, Result, Context};
+use heck::{ToShoutySnakeCase, ToUpperCamelCase};
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+use crate::codegen::Codegen;
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct CwtTemplate {
+    /// Name of the certificate.
+    pub name: String,
+    /// Variable declarations.
+    pub variables: IndexMap<String, VariableType>,
+    /// Variable declarations.
+    pub constants: IndexMap<String, ConstantType>,
+    /// Certificate specification.
+    pub certificate: StructureType,
+}
+
+/// Declaration of a variable that can be filled into the template.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(tag = "type", rename_all = "kebab-case")]
+pub enum VariableType {
+    /// Raw array of bytes.
+    ByteArray {
+        /// Length in bytes for this variable.
+        size: usize,
+    },
+    /// Integer.
+    Integer,
+    /// UTF-8 encoded String.
+    String {
+        /// Maximum size in bytes for this variable.
+        size: usize,
+    },
+}
+
+impl VariableType {
+    fn get_max_size(&self) -> usize {
+        match self {
+            VariableType::ByteArray{size} => 1 + 8 + size,
+            VariableType::Integer => 1 + 8,
+            VariableType::String{size} => 1 + 8 + size,
+        }
+    }
+}
+
+/// Declaration of a variable that can be filled into the template.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(tag = "type", rename_all = "kebab-case")]
+pub enum ConstantType {
+    /// Raw array of bytes.
+    ByteArray {
+        /// Big-endian hex-encoded bytes
+        #[serde(with = "hex::serde")]
+        value: Vec<u8>,
+    },
+    /// Integer.
+    Integer {
+        value: i64,
+    },
+    /// UTF-8 encoded String.
+    String {
+        value: String,
+    },
+}
+
+impl ConstantType {
+    fn get_max_size(&self) -> usize {
+        match self {
+            ConstantType::ByteArray{value} => 1 + 8 + value.len(),
+            ConstantType::Integer{..} => 1 + 8,
+            ConstantType::String{value} => 1 + 8 + value.len(),
+        }
+    }
+}
+
+/// Declaration of a variable that can be filled into the template.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum StructureType {
+    Item(String),
+    Map(IndexMap<String, StructureType>),
+    Array(Vec<StructureType>),
+}
+
+impl StructureType {
+    fn get_max_size(&self, variables: &IndexMap<String, VariableType>, constants: &IndexMap<String, ConstantType>) -> Result<usize> {
+        let lookup = |var_name| {
+            if let Some(var) = variables.get(var_name) {
+                return Ok(var.get_max_size())
+            }
+
+            if let Some(var) = constants.get(var_name) {
+                return Ok(var.get_max_size())
+            }
+
+            return Err(anyhow!("Cannot find {} in variables or constants", var_name));
```

Comment:
you can use `bail!()` from `anyhow` which does the same thing but which I find a bit clearer